static/app/components/tooltip.tsx

```diff
@@ -116,215 +111,262 @@ function computeOriginFromArrow(
+ * with special care to unregister tooltips if they become ignored.
+ */
+function useTooltipInAcceptanceTests({
+  tooltipId,
+  setOpen,
+  setUseGlobalPortal,
+  ignore,
+}: UseTooltipInAcceptanceTestsOtpions) {
+  if (!IS_ACCEPTANCE_TEST) {
+    return;
   }
 
-  tooltipId: string = domId('tooltip-');
-  delayTimeout: number | null = null;
-  delayHideTimeout: number | null = null;
-  triggerEl: Element | null = null;
-
-  getPortal = memoize((usesGlobalPortal): HTMLElement => {
-    if (usesGlobalPortal) {
-      let portal = document.getElementById('tooltip-portal');
-      if (!portal) {
-        portal = document.createElement('div');
-        portal.setAttribute('id', 'tooltip-portal');
-        document.body.appendChild(portal);
-      }
-      return portal;
-    }
-    const portal = document.createElement('div');
-    document.body.appendChild(portal);
-    return portal;
-  });
+  async function registerTooltip() {
+    const tooltipStore = await getAcceptanceTestTooltipStore();
+    tooltipStore?.addTooltip(tooltipId, {setOpen, setUseGlobalPortal});
+  }
 
-  setOpen = () => {
-    this.setState({isOpen: true});
-  };
+  async function unregisterTooltip() {
+    const tooltipStore = await getAcceptanceTestTooltipStore();
+    tooltipStore?.removeTooltip(tooltipId);
+  }
 
-  setClose = () => {
-    this.setState({isOpen: false});
-  };
+  // Register and unregister tooltips from the store
+  useEffect(() => {
+    if (!ignore) {
+      registerTooltip();
+    }
+    return () => void unregisterTooltip();
+  }, [ignore]);
+}
 
-  handleOpen = () => {
-    const {delay, showOnlyOnOverflow} = this.props;
+function Tooltip({
+  children,
+  className,
+  delay,
+  forceShow,
+  isHoverable,
+  popperStyle,
+  showOnlyOnOverflow,
+  skipWrapper,
+  title,
+  disabled = false,
+  disableForVisualTest = false,
+  position = 'top',
+  containerDisplayMode = 'inline-block',
+}: Props) {
+  const [isOpen, setOpen] = useState(false);
+
+  // XXX: We use this when running acceptance tests with 'open all tooltips'
+  const [useGlobalPortal, setUseGlobalPortal] = useState(true);
+
+  // Tooltip ID is stable accross renders
+  const tooltipId = useMemo(() => domId('tooltip-'), []);
+
+  // Delayed open and close time handles
+  const delayOpenTimeout = useRef<number | null>(null);
+  const delayHideTimeout = useRef<number | null>(null);
+
+  // Tracks the triggering element
+  const triggerEl = useRef<HTMLElement | null>(null);
+
+  const tooltipPortal = useMemo(
+    () => getTooltipPortal(useGlobalPortal),
+    [useGlobalPortal]
+  );
+
+  // When not using a global portal we need to be sure to cleanup tooltip
+  // portals from the DOM
+  useEffect(
+    () => () => void !useGlobalPortal && document.body.removeChild(tooltipPortal),
+    [useGlobalPortal, tooltipPortal]
+  );
+
+  // Register functionality for visual acceptance tests
+  useTooltipInAcceptanceTests({
```

Comment:
I'm not sure if hooks are the best way to go here as we cant conditionally call them (obviously I like the separation of concern), but we should be able to strip out more dead code if we write this as a function and call it like
```
useEffect(() => {
  if(IS_ACCEPTANCE_TEST) {
    registerTooltipForAcceptanceTest()
  }
},[])
```

Reply:
well the hook is already a no-op immediately once the IS_ACCEPTANCE_TEST is false right?

Reply:
Yes, correct. But what you are doing is actually invalid React code because you are calling conditionally calling useEffect due to that return statement. Idk why our linting doesnt warn you, but you should be seeing this error. ([example](https://codesandbox.io/s/eloquent-rumple-ils942?file=/src/App.tsx))
```
React Hook "useEffect" is called conditionally. React Hooks must be called in the exact same order in every component render. Did you accidentally call a React Hook after an early return? (react-hooks/rules-of-hooks)eslint
```

I think eod, even if hooks were to support conditional calls, we should still opt to use function calls for this because they signal side effect behavior more cleanly.

Reply:
Yes but this constant will never change values across renders :)

Reply:
There's no real static variables in JS, so there's never a guaranteed that the variable wont change. The linter will complain in the future and this will error, so we'll need to refactor it to a fn anyways. Our linter config currently has no react hook rules, so once we add those, this will have to be refactored. Even though you are correct, this is a pattern we should not encourage because it's invalid react. I dont want someone to introduce a similar pattern in their component because right now it will pass CI and all checks, but it will crash at runtime.

Reply:
I'll change it over in a PR right after this. Been trying to merge this all week lol

Reply:
I think we should never break hook rules or prefer cosmetics over correctness. Eod, the hook you added already has a side effect, it's just in a different place.

The better way to do this would be to create a wrapper component that would either render a AcceptanceTestTooltip (which renders tooltip + exposes acceptance test functionality) vs just rendering the regular Tooltip based off IS_ACCEPTANCE_TEST variable. This way you would keep the code isolated between the two and could enhance the tooltip with this functionality. 

I dont think refactoring this to a fn call is a lot of work in this case, but since @scttcper approved this, I'll let you judge that :)

Reply:
Will follow up with an update.

Reply:
Just my $0.02, but I also lean towards following hook rules over an extra `useEffect` in prod, the hook rules usually exist to help reduce number of side effects, which are usually the most annoying errors to debug and fix. They can be a real pain...

Incurring a few μs for an if + useEffect isn't the best and def feels meh, but if a better solution doesn't exist or doesn't seem worth the effort, I wouldn't leave a potential side-effect source in the code-base, even assuming it has no real world effect. Just considering the pattern implications alone (someone might see it an unknowingly copy some of the behavior, etc.) it doesn't seem worth it.